### DISCUSSION TITLE
Fuent API and Logback TurboFilter

### DISCUSSION THREAD

**Discussion by Kiiv** [qos-ch/slf4j#343] [Q&A]
Hi,

don't know if it's more a Logback issue or an slf4j one.

Logback offers to implement TurboFilters (https://logback.qos.ch/manual/filters.html#TurboFilter).

If I'm logging something with the old API just like :
```java
   logger.error("error happened", new RuntimeException("Hi"));
```
then slf4j just call the error method of the underlying logger implementation. 
With Logback it's something like :

```java
    private void filterAndLog_2(final String localFQCN, final Marker marker, final Level level, final String msg,
            final Object param1, final Object param2, final Throwable t) {

        final FilterReply decision = loggerContext.getTurboFilterChainDecision_2(marker, this, level, msg, param1,
                param2, t);
        ...
    }

```
As you can see, the turboFilterChain is called with all the information passed to the method.

But if I'm logging something with the fluent API juste like :
```java
logger.atError().setCause(new RuntimeException("Hi")).log("error happened");
```
then slf4j call the 'log' method of the underlying logger implementation. 
This implementation completely bypass the TurboFilter chain.

The TurboFilter is still called when the "atError" method is called because the 'isErrorEnabled' method call it but at this step there is no information available about the message or exception to log.

To conclude, the fluent API does not seem really backward compatible, at least with Logback :/

**Comment by ceki** [comment]
This issue is closely related to [logback/871](https://github.com/qos-ch/logback/issues/871). I am closing this discussion as a duplicate.
